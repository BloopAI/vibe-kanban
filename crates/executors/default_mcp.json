{
  "vibe_kanban": {
    "command": "npx",
    "args": [
      "-y",
      "automagik-forge",
      "--mcp"
    ]
  },
  "context7": {
    "type": "http",
    "url": "https://mcp.context7.com/mcp",
    "headers": {
      "CONTEXT7_API_KEY": "YOUR_API_KEY"
    }
  },
  "playwright": {
    "command": "npx",
    "args": [
      "@playwright/mcp@latest"
    ]
  },
  "meta": {
    "vibe_kanban": {
      "name": "Automagik Forge",
      "description": "Create, update and delete Automagik Forge tasks",
      "url": "https://forge.automag.ik/docs/integrations/automagik-forge-mcp-server",
      "icon": "forge-clear.svg"
    },
    "context7": {
      "name": "Context7",
      "description": "Fetch up-to-date documentation and code examples",
      "url": "https://github.com/upstash/context7",
      "icon": "mcp/context7logo.png"
    },
    "playwright": {
      "name": "Playwright",
      "description": "Browser automation with Playwright",
      "url": "https://github.com/microsoft/playwright-mcp",
      "icon": "mcp/playwright_logo_icon.svg"
    }
  }
}
