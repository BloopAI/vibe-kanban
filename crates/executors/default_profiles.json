{
  "profiles": {
    "claude-code": {
      "label": "claude-code",
      "mcp_config_path": null,
      "CLAUDE_CODE": {
        "variant": "claude_code",
        "plan": false,
        "append_prompt": null
      },
      "variants": {
        "plan": {
          "mcp_config_path": null,
          "CLAUDE_CODE": {
            "variant": "claude_code",
            "plan": true,
            "append_prompt": null
          }
        }
      }
    },
    "claude-code-router": {
      "label": "claude-code-router",
      "mcp_config_path": null,
      "CLAUDE_CODE": {
        "variant": "claude_code_router",
        "plan": false,
        "append_prompt": null
      },
      "variants": {}
    },
    "amp": {
      "label": "amp",
      "mcp_config_path": null,
      "AMP": {
        "append_prompt": null
      },
      "variants": {}
    },
    "gemini": {
      "label": "gemini",
      "mcp_config_path": null,
      "GEMINI": {
        "model": "default",
        "append_prompt": null
      },
      "variants": {
        "flash": {
          "mcp_config_path": null,
          "GEMINI": {
            "model": "flash",
            "append_prompt": null
          }
        }
      }
    },
    "codex": {
      "label": "codex",
      "mcp_config_path": null,
      "CODEX": {
        "append_prompt": null
      },
      "variants": {}
    },
    "opencode": {
      "label": "opencode",
      "mcp_config_path": null,
      "OPENCODE": {
        "append_prompt": null
      },
      "variants": {}
    },
    "qwen-code": {
      "label": "qwen-code",
      "mcp_config_path": "~/.qwen/settings.json",
      "QWEN_CODE": {
        "append_prompt": null
      },
      "variants": {}
    },
    "cursor": {
      "label": "cursor",
      "mcp_config_path": null,
      "CURSOR": {
        "append_prompt": null
      },
      "variants": {}
    }
  }
}