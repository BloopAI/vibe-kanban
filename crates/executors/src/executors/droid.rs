use std::{collections::HashMap, path::Path, process::Stdio, sync::Arc};

use async_trait::async_trait;
use command_group::AsyncCommandGroup;
use futures::StreamExt;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use tokio::{io::AsyncWriteExt, process::Command};
use ts_rs::TS;
use workspace_utils::{log_msg::LogMsg, msg_store::MsgStore, shell::get_shell_command};

use crate::{
    command::{CmdOverrides, CommandBuilder, apply_overrides},
    executors::{AppendPrompt, ExecutorError, SpawnedChild, StandardCodingAgentExecutor},
    logs::{
        ActionType, CommandExitStatus, CommandRunResult, NormalizedEntry, NormalizedEntryType,
        TodoItem, ToolStatus,
        stderr_processor::normalize_stderr_logs,
        utils::{EntryIndexProvider, patch::ConversationPatch},
    },
};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, TS, JsonSchema)]
#[serde(rename_all = "kebab-case")]
pub enum Autonomy {
    Normal,
    Low,
    Medium,
    High,
    SkipPermissionsUnsafe,
}

fn default_autonomy() -> Autonomy {
    Autonomy::Medium
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, TS, JsonSchema)]
pub struct Droid {
    #[serde(default)]
    pub append_prompt: AppendPrompt,

    #[serde(default = "default_autonomy")]
    #[schemars(
        title = "Autonomy Level",
        description = "Permission level for file and system operations"
    )]
    pub autonomy: Autonomy,

    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[schemars(
        title = "Model",
        description = "Model to use (e.g., gpt-5-codex, claude-sonnet-4)"
    )]
    pub model: Option<String>,

    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[schemars(
        title = "Reasoning Effort",
        description = "Reasoning effort level: off, low, medium, high"
    )]
    pub reasoning_effort: Option<String>,

    #[serde(flatten)]
    pub cmd: CmdOverrides,
}

impl Droid {
    fn build_command_builder(&self) -> CommandBuilder {
        let mut builder = CommandBuilder::new("droid exec").params(["--output-format=stream-json"]);

        let autonomy_args: Vec<&str> = match &self.autonomy {
            Autonomy::Normal => vec![],
            Autonomy::Low => vec!["--auto", "low"],
            Autonomy::Medium => vec!["--auto", "medium"],
            Autonomy::High => vec!["--auto", "high"],
            Autonomy::SkipPermissionsUnsafe => vec!["--skip-permissions-unsafe"],
        };
        builder = builder.extend_params(autonomy_args);

        if let Some(ref model) = self.model {
            builder = builder.extend_params(["--model", model]);
        }

        if let Some(ref effort) = self.reasoning_effort {
            builder = builder.extend_params(["--reasoning-effort", effort]);
        }

        apply_overrides(builder, &self.cmd)
    }
}

#[async_trait]
impl StandardCodingAgentExecutor for Droid {
    async fn spawn(&self, current_dir: &Path, prompt: &str) -> Result<SpawnedChild, ExecutorError> {
        let (shell_cmd, shell_arg) = get_shell_command();
        let droid_command = self.build_command_builder().build_initial();
        let combined_prompt = self.append_prompt.combine_prompt(prompt);

        let mut command = Command::new(shell_cmd);
        command
            .kill_on_drop(true)
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .current_dir(current_dir)
            .arg(shell_arg)
            .arg(&droid_command);

        let mut child = command.group_spawn()?;

        // Feed the prompt in, then close the pipe so droid sees EOF
        if let Some(mut stdin) = child.inner().stdin.take() {
            stdin.write_all(combined_prompt.as_bytes()).await?;
            stdin.shutdown().await?;
        }

        Ok(child.into())
    }

    async fn spawn_follow_up(
        &self,
        current_dir: &Path,
        prompt: &str,
        session_id: &str,
    ) -> Result<SpawnedChild, ExecutorError> {
        // Use shell command for cross-platform compatibility
        let (shell_cmd, shell_arg) = get_shell_command();
        let continue_cmd = self
            .build_command_builder()
            .build_follow_up(&["--session-id".to_string(), session_id.to_string()]);

        let combined_prompt = self.append_prompt.combine_prompt(prompt);

        let mut command = Command::new(shell_cmd);
        command
            .kill_on_drop(true)
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .current_dir(current_dir)
            .arg(shell_arg)
            .arg(&continue_cmd);

        let mut child = command.group_spawn()?;

        // Feed the prompt in, then close the pipe so droid sees EOF
        if let Some(mut stdin) = child.inner().stdin.take() {
            stdin.write_all(combined_prompt.as_bytes()).await?;
            stdin.shutdown().await?;
        }

        Ok(child.into())
    }

    fn normalize_logs(&self, msg_store: Arc<MsgStore>, current_dir: &Path) {
        let entry_index_provider = EntryIndexProvider::start_from(&msg_store);

        DroidLogProcessor::process_logs(
            msg_store.clone(),
            current_dir,
            entry_index_provider.clone(),
        );

        // Process stderr logs using the standard stderr processor
        normalize_stderr_logs(msg_store, entry_index_provider);
    }

    // MCP configuration methods
    fn default_mcp_config_path(&self) -> Option<std::path::PathBuf> {
        dirs::home_dir().map(|home| home.join(".factory").join("mcp.json"))
    }
}

#[derive(Deserialize, Serialize, Debug, Clone)]
#[serde(tag = "type", rename_all = "snake_case")]
enum DroidJson {
    System {
        #[serde(default)]
        subtype: Option<String>,
        session_id: String,
        #[serde(default)]
        cwd: Option<String>,
        #[serde(default)]
        tools: Option<Vec<String>>,
        #[serde(default)]
        model: Option<String>,
    },
    Message {
        role: String,
        id: String,
        text: String,
        timestamp: u64,
        session_id: String,
    },
    ToolCall {
        id: String,
        #[serde(rename = "messageId")]
        message_id: String,
        #[serde(rename = "toolId")]
        tool_id: String,
        #[serde(rename = "toolName")]
        tool_name: String,
        parameters: serde_json::Value,
        timestamp: u64,
        session_id: String,
    },
    ToolResult {
        id: String,
        #[serde(rename = "messageId")]
        message_id: String,
        #[serde(rename = "toolId")]
        tool_id: String,
        #[serde(rename = "isError")]
        is_error: bool,
        value: serde_json::Value,
        timestamp: u64,
        session_id: String,
    },
    Error {
        source: String,
        message: String,
        timestamp: u64,
    },
}

struct ToolCallInfo {
    tool_name: String,
    entry_index: usize,
    action_type: ActionType,
    content: String,
}

struct DroidLogProcessor {
    tool_map: HashMap<String, ToolCallInfo>,
}

impl DroidLogProcessor {
    fn process_logs(
        msg_store: Arc<MsgStore>,
        _current_dir: &Path,
        entry_index_provider: EntryIndexProvider,
    ) {
        tokio::spawn(async move {
            let mut stream = msg_store.history_plus_stream();
            let mut buffer = String::new();
            let mut session_id_extracted = false;
            let mut processor = Self {
                tool_map: HashMap::new(),
            };

            while let Some(Ok(msg)) = stream.next().await {
                let chunk = match msg {
                    LogMsg::Stdout(x) => x,
                    LogMsg::JsonPatch(_) | LogMsg::SessionId(_) | LogMsg::Stderr(_) => continue,
                    LogMsg::Finished => break,
                };

                buffer.push_str(&chunk);

                for line in buffer
                    .split_inclusive('\n')
                    .filter(|l| l.ends_with('\n'))
                    .map(str::to_owned)
                    .collect::<Vec<_>>()
                {
                    let trimmed = line.trim();
                    if trimmed.is_empty() {
                        continue;
                    }

                    match serde_json::from_str::<DroidJson>(trimmed) {
                        Ok(droid_json) => {
                            if !session_id_extracted
                                && let Some(session_id) = Self::extract_session_id(&droid_json)
                            {
                                msg_store.push_session_id(session_id);
                                session_id_extracted = true;
                            }

                            let patches =
                                processor.normalize_entries(&droid_json, &entry_index_provider);
                            for patch in patches {
                                msg_store.push_patch(patch);
                            }
                        }
                        Err(_) => {
                            if !trimmed.is_empty() {
                                let entry = NormalizedEntry {
                                    timestamp: None,
                                    entry_type: NormalizedEntryType::SystemMessage,
                                    content: trimmed.to_string(),
                                    metadata: None,
                                };

                                let patch_id = entry_index_provider.next();
                                let patch =
                                    ConversationPatch::add_normalized_entry(patch_id, entry);
                                msg_store.push_patch(patch);
                            }
                        }
                    }
                }

                buffer = buffer.rsplit('\n').next().unwrap_or("").to_owned();
            }

            if !buffer.trim().is_empty() {
                let entry = NormalizedEntry {
                    timestamp: None,
                    entry_type: NormalizedEntryType::SystemMessage,
                    content: buffer.trim().to_string(),
                    metadata: None,
                };

                let patch_id = entry_index_provider.next();
                let patch = ConversationPatch::add_normalized_entry(patch_id, entry);
                msg_store.push_patch(patch);
            }
        });
    }

    fn extract_session_id(json: &DroidJson) -> Option<String> {
        match json {
            DroidJson::System { session_id, .. } => Some(session_id.clone()),
            DroidJson::Message { session_id, .. } => Some(session_id.clone()),
            DroidJson::ToolCall { session_id, .. } => Some(session_id.clone()),
            DroidJson::ToolResult { session_id, .. } => Some(session_id.clone()),
            DroidJson::Error { .. } => None,
        }
    }

    fn normalize_entries(
        &mut self,
        json: &DroidJson,
        entry_index_provider: &EntryIndexProvider,
    ) -> Vec<json_patch::Patch> {
        match json {
            DroidJson::System { .. } => {
                vec![]
            }
            DroidJson::Message { role, text, .. } => {
                let entry_type = match role.as_str() {
                    "user" => NormalizedEntryType::UserMessage,
                    "assistant" => NormalizedEntryType::AssistantMessage,
                    _ => NormalizedEntryType::SystemMessage,
                };
                vec![ConversationPatch::add_normalized_entry(
                    entry_index_provider.next(),
                    NormalizedEntry {
                        timestamp: None,
                        entry_type,
                        content: text.clone(),
                        metadata: None,
                    },
                )]
            }
            DroidJson::ToolCall {
                tool_name,
                parameters,
                id,
                ..
            } => {
                let action_type = Self::map_tool_to_action(tool_name, parameters);
                let content = Self::generate_concise_content(tool_name, &action_type);
                let entry_idx = entry_index_provider.next();

                self.tool_map.insert(
                    id.clone(),
                    ToolCallInfo {
                        tool_name: tool_name.clone(),
                        entry_index: entry_idx,
                        action_type: action_type.clone(),
                        content: content.clone(),
                    },
                );

                vec![ConversationPatch::add_normalized_entry(
                    entry_idx,
                    NormalizedEntry {
                        timestamp: None,
                        entry_type: NormalizedEntryType::ToolUse {
                            tool_name: tool_name.clone(),
                            action_type,
                            status: ToolStatus::Created,
                        },
                        content,
                        metadata: Some(
                            serde_json::to_value(parameters).unwrap_or(serde_json::Value::Null),
                        ),
                    },
                )]
            }
            DroidJson::ToolResult {
                id,
                is_error,
                value,
                ..
            } => {
                if let Some(info) = self.tool_map.get(id) {
                    let status = if *is_error {
                        ToolStatus::Failed
                    } else {
                        ToolStatus::Success
                    };

                    // Parse the result value
                    let result_str = if let Some(s) = value.as_str() {
                        s.to_string()
                    } else {
                        serde_json::to_string_pretty(value).unwrap_or_default()
                    };

                    // For Execute commands, parse the result and extract exit code
                    let updated_action_type =
                        if matches!(info.action_type, ActionType::CommandRun { .. }) {
                            // Try to extract exit code from "[Process exited with code X]" pattern
                            let exit_code = result_str
                                .lines()
                                .find(|line| line.contains("[Process exited with code"))
                                .and_then(|line| {
                                    line.strip_prefix("[Process exited with code ")?
                                        .strip_suffix("]")?
                                        .parse::<i32>()
                                        .ok()
                                });

                            let result = Some(CommandRunResult {
                                exit_status: exit_code
                                    .map(|code| CommandExitStatus::ExitCode { code })
                                    .or(Some(CommandExitStatus::Success { success: !is_error })),
                                output: Some(result_str.clone()),
                            });

                            if let ActionType::CommandRun { command, .. } = &info.action_type {
                                ActionType::CommandRun {
                                    command: command.clone(),
                                    result,
                                }
                            } else {
                                info.action_type.clone()
                            }
                        } else {
                            info.action_type.clone()
                        };

                    // Create updated entry with status and result
                    let entry = NormalizedEntry {
                        timestamp: None,
                        entry_type: NormalizedEntryType::ToolUse {
                            tool_name: info.tool_name.clone(),
                            action_type: updated_action_type,
                            status,
                        },
                        content: info.content.clone(),
                        metadata: None,
                    };

                    vec![ConversationPatch::replace(info.entry_index, entry)]
                } else {
                    vec![]
                }
            }
            DroidJson::Error { message, .. } => {
                vec![ConversationPatch::add_normalized_entry(
                    entry_index_provider.next(),
                    NormalizedEntry {
                        timestamp: None,
                        entry_type: NormalizedEntryType::ErrorMessage,
                        content: message.clone(),
                        metadata: None,
                    },
                )]
            }
        }
    }

    fn generate_concise_content(tool_name: &str, action_type: &ActionType) -> String {
        match action_type {
            ActionType::FileRead { path } => format!("`{path}`"),
            ActionType::FileEdit { path, .. } => format!("`{path}`"),
            ActionType::CommandRun { command, .. } => format!("`{command}`"),
            ActionType::Search { query } => format!("`{query}`"),
            ActionType::WebFetch { url } => format!("`{url}`"),
            ActionType::TodoManagement { todos, .. } => {
                if todos.is_empty() {
                    "Todo list".to_string()
                } else {
                    format!("{} todo items", todos.len())
                }
            }
            ActionType::Other { description } => description.clone(),
            _ => tool_name.to_string(),
        }
    }

    fn map_tool_to_action(tool_name: &str, params: &serde_json::Value) -> ActionType {
        match tool_name {
            "Read" => ActionType::FileRead {
                path: params
                    .get("file_path")
                    .or_else(|| params.get("path"))
                    .and_then(|v| v.as_str())
                    .unwrap_or("")
                    .to_string(),
            },
            "LS" => ActionType::FileRead {
                path: params
                    .get("directory_path")
                    .or_else(|| params.get("path"))
                    .and_then(|v| v.as_str())
                    .unwrap_or("")
                    .to_string(),
            },
            "Glob" => ActionType::FileRead {
                path: params
                    .get("folder")
                    .or_else(|| params.get("path"))
                    .and_then(|v| v.as_str())
                    .unwrap_or("")
                    .to_string(),
            },
            "Grep" => ActionType::FileRead {
                path: params
                    .get("path")
                    .and_then(|v| v.as_str())
                    .unwrap_or("")
                    .to_string(),
            },
            "Execute" => ActionType::CommandRun {
                command: params
                    .get("command")
                    .and_then(|v| v.as_str())
                    .unwrap_or("")
                    .to_string(),
                result: None,
            },
            "Edit" | "MultiEdit" | "Create" => ActionType::FileEdit {
                path: params
                    .get("file_path")
                    .or_else(|| params.get("path"))
                    .and_then(|v| v.as_str())
                    .unwrap_or("")
                    .to_string(),
                changes: vec![],
            },
            "ApplyPatch" => ActionType::FileEdit {
                path: Self::extract_path_from_patch(params),
                changes: vec![],
            },
            "TodoWrite" => {
                let todos = params
                    .get("todos")
                    .and_then(|v| v.as_array())
                    .map(|arr| {
                        arr.iter()
                            .filter_map(|item| {
                                Some(TodoItem {
                                    content: item.get("content")?.as_str()?.to_string(),
                                    status: item.get("status")?.as_str()?.to_string(),
                                    priority: item
                                        .get("priority")
                                        .and_then(|p| p.as_str())
                                        .map(|s| s.to_string()),
                                })
                            })
                            .collect()
                    })
                    .unwrap_or_default();

                ActionType::TodoManagement {
                    todos,
                    operation: "update".to_string(),
                }
            }
            "WebSearch" => ActionType::WebFetch {
                url: params
                    .get("query")
                    .and_then(|v| v.as_str())
                    .unwrap_or("")
                    .to_string(),
            },
            "FetchUrl" => ActionType::WebFetch {
                url: params
                    .get("url")
                    .and_then(|v| v.as_str())
                    .unwrap_or("")
                    .to_string(),
            },
            _ => ActionType::Other {
                description: tool_name.to_string(),
            },
        }
    }

    fn extract_path_from_patch(params: &serde_json::Value) -> String {
        if let Some(input) = params.get("input").and_then(|v| v.as_str()) {
            for line in input.lines() {
                if line.starts_with("*** Update File:") || line.starts_with("*** Create File:") {
                    return line
                        .split(':')
                        .nth(1)
                        .map(|s| s.trim().to_string())
                        .unwrap_or_default();
                }
            }
        }
        String::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_droid_json_parsing() {
        let system_msg = r#"{"type":"system","subtype":"init","cwd":"/test","session_id":"test-123","tools":["Read"],"model":"gpt-5-codex"}"#;
        let user_msg = r#"{"type":"message","role":"user","id":"u1","text":"hello","timestamp":12345,"session_id":"test-123"}"#;
        let tool_call = r#"{"type":"tool_call","id":"t1","messageId":"m1","toolId":"Read","toolName":"Read","parameters":{"file_path":"test.txt"},"timestamp":12345,"session_id":"test-123"}"#;

        let parsed_system: Result<DroidJson, _> = serde_json::from_str(system_msg);
        let parsed_user: Result<DroidJson, _> = serde_json::from_str(user_msg);
        let parsed_tool: Result<DroidJson, _> = serde_json::from_str(tool_call);

        assert!(parsed_system.is_ok(), "System message should parse");
        assert!(parsed_user.is_ok(), "User message should parse");
        assert!(parsed_tool.is_ok(), "Tool call should parse");
    }
}
