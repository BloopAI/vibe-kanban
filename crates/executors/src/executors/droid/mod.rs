use std::{path::Path, process::Stdio, sync::Arc};

use async_trait::async_trait;
use command_group::AsyncCommandGroup;
use tokio::{io::AsyncWriteExt, process::Command};
use workspace_utils::{msg_store::MsgStore, shell::get_shell_command};

use crate::{
    executors::{ExecutorError, SpawnedChild, StandardCodingAgentExecutor},
    logs::{stderr_processor::normalize_stderr_logs, utils::EntryIndexProvider},
};

mod action_mapper;
mod events;
mod patch_emitter;
mod processor;
mod types;

use processor::DroidLogProcessor;
pub use types::{Autonomy, Droid};

async fn exec_command_with_prompt(
    cmd: &String,
    prompt: &String,
    current_dir: &Path,
) -> Result<SpawnedChild, ExecutorError> {
    let (shell_cmd, shell_arg) = get_shell_command();
    let mut command = Command::new(shell_cmd);
    command
        .kill_on_drop(true)
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .current_dir(current_dir)
        .arg(shell_arg)
        .arg(&cmd);

    let mut child = command.group_spawn()?;

    if let Some(mut stdin) = child.inner().stdin.take() {
        stdin.write_all(prompt.as_bytes()).await?;
        stdin.shutdown().await?;
    }

    Ok(child.into())
}

#[async_trait]
impl StandardCodingAgentExecutor for Droid {
    async fn spawn(&self, current_dir: &Path, prompt: &str) -> Result<SpawnedChild, ExecutorError> {
        let droid_command = self.build_command_builder().build_initial();
        let combined_prompt = self.append_prompt.combine_prompt(prompt);

        exec_command_with_prompt(&droid_command, &combined_prompt, current_dir).await
    }

    async fn spawn_follow_up(
        &self,
        current_dir: &Path,
        prompt: &str,
        session_id: &str,
    ) -> Result<SpawnedChild, ExecutorError> {
        let continue_cmd = self
            .build_command_builder()
            .build_follow_up(&["--session-id".to_string(), session_id.to_string()]);
        let combined_prompt = self.append_prompt.combine_prompt(prompt);

        exec_command_with_prompt(&continue_cmd, &combined_prompt, current_dir).await
    }

    fn normalize_logs(&self, msg_store: Arc<MsgStore>, current_dir: &Path) {
        let entry_index_provider = EntryIndexProvider::start_from(&msg_store);
        DroidLogProcessor::process_logs(
            msg_store.clone(),
            current_dir,
            entry_index_provider.clone(),
        );

        normalize_stderr_logs(msg_store, entry_index_provider);
    }

    fn default_mcp_config_path(&self) -> Option<std::path::PathBuf> {
        dirs::home_dir().map(|home| home.join(".factory").join("mcp.json"))
    }
}
