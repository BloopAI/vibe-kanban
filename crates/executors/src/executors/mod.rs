use std::path::PathBuf;

use async_trait::async_trait;
use command_group::AsyncGroupChild;
use futures_io::Error as FuturesIoError;
use thiserror::Error;
use uuid::Uuid;

use crate::logs::NormalizedConversation;

pub mod standard;

#[derive(Debug, Error)]
pub enum ExecutorError {
    #[error("Follow-up is not supported")]
    FollowUpNotSupported,
    #[error(transparent)]
    SpawnError(#[from] FuturesIoError),
}

/// Trait for coding agents that can execute tasks, normalize logs, and support follow-up sessions
#[async_trait]
pub trait Executor: Send + Sync {
    async fn spawn(&self, current_dir: &PathBuf) -> Result<AsyncGroupChild, ExecutorError>;
    async fn spawn_followup(
        &self,
        current_dir: &PathBuf,
    ) -> Result<AsyncGroupChild, ExecutorError> {
        Err(ExecutorError::FollowUpNotSupported)
    }
}
