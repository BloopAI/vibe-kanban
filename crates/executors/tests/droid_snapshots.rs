use std::{fs, path::Path};

use executors::executors::droid::{
    events::{LogEvent, ProcessorState, process_event},
    types::DroidJson,
};

#[derive(serde::Serialize)]
struct SessionSnapshot {
    final_state: ProcessorState,
    events: Vec<LogEvent>,
    event_count: usize,
}

fn process_jsonl_file(file_path: &str) -> SessionSnapshot {
    let content = fs::read_to_string(file_path).expect("Failed to read test file");
    let worktree_path = Path::new("/tmp/test-worktree");

    let mut state = ProcessorState::default();
    let mut all_events = Vec::new();

    for line in content.lines() {
        let line = line.trim();
        if line.is_empty() {
            continue;
        }

        let event: DroidJson = match serde_json::from_str(line) {
            Ok(e) => e,
            Err(e) => {
                eprintln!("Failed to parse line: {line}\nError: {e}");
                continue;
            }
        };

        let events = process_event(&mut state, &event, worktree_path);
        all_events.extend(events);
    }

    SessionSnapshot {
        event_count: all_events.len(),
        events: all_events,
        final_state: state,
    }
}

#[test]
fn test_hello_world_session() {
    let snapshot = process_jsonl_file("tests/droid_snapshots/fixtures/hello-world.jsonl");
    insta::assert_yaml_snapshot!(snapshot);
}

#[test]
fn test_edits_and_execution_session() {
    let snapshot = process_jsonl_file("tests/droid_snapshots/fixtures/edits-and-execution.jsonl");
    insta::assert_yaml_snapshot!(snapshot);
}

#[test]
fn test_glob_permission_denied_session() {
    let snapshot = process_jsonl_file("tests/droid_snapshots/fixtures/glob.jsonl");
    insta::assert_yaml_snapshot!(snapshot);
}

#[test]
fn test_insufficient_perms_session() {
    let snapshot = process_jsonl_file("tests/droid_snapshots/fixtures/insufficient-perms.jsonl");
    insta::assert_yaml_snapshot!(snapshot);
}
