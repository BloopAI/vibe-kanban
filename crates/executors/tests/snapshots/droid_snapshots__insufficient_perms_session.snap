---
source: crates/executors/tests/droid_snapshots.rs
expression: snapshot
---
log_event_converter:
  tool_map: {}
  model_reported: true
  pending_fifo: []
events:
  - AddEntry:
      timestamp: ~
      entry_type:
        type: system_message
      content: "model: gpt-5-codex"
      metadata: ~
  - AddEntry:
      timestamp: ~
      entry_type:
        type: user_message
      content: "create a text file called 'a.txt' with the character 'a' inside"
      metadata: ~
  - AddToolCall:
      tool_call_id: call_WaiIpWDOVLcWkCDEJKogOzrl
      entry:
        timestamp: ~
        entry_type:
          type: tool_use
          tool_name: Read
          action_type:
            action: file_read
            path: /Users/tempuser/projects/vibe-kanban/README.md
          status:
            status: created
        content: "`/Users/tempuser/projects/vibe-kanban/README.md`"
        metadata: ~
  - UpdateToolCall:
      tool_call_id: call_WaiIpWDOVLcWkCDEJKogOzrl
      entry:
        timestamp: ~
        entry_type:
          type: tool_use
          tool_name: Read
          action_type:
            action: file_read
            path: /Users/tempuser/projects/vibe-kanban/README.md
          status:
            status: success
        content: "`/Users/tempuser/projects/vibe-kanban/README.md`"
        metadata: ~
  - AddToolCall:
      tool_call_id: call_ULikm3YyJS9DBwHRcEck7ejc
      entry:
        timestamp: ~
        entry_type:
          type: tool_use
          tool_name: TodoWrite
          action_type:
            action: todo_management
            todos:
              - content: "Create text file a.txt containing single character 'a'"
                status: in_progress
                priority: medium
            operation: update
          status:
            status: created
        content: TODO list updated
        metadata: ~
  - UpdateToolCall:
      tool_call_id: call_ULikm3YyJS9DBwHRcEck7ejc
      entry:
        timestamp: ~
        entry_type:
          type: tool_use
          tool_name: TodoWrite
          action_type:
            action: todo_management
            todos:
              - content: "Create text file a.txt containing single character 'a'"
                status: in_progress
                priority: medium
            operation: update
          status:
            status: success
        content: TODO list updated
        metadata: ~
  - AddToolCall:
      tool_call_id: call_W4t4W4TDhrZFiM9DxVox21PW
      entry:
        timestamp: ~
        entry_type:
          type: tool_use
          tool_name: Execute
          action_type:
            action: command_run
            command: "printf 'a' > \"/Users/tempuser/projects/vibe-kanban/a.txt\""
            result: ~
          status:
            status: created
        content: "`printf 'a' > \"/Users/tempuser/projects/vibe-kanban/a.txt\"`"
        metadata: ~
  - UpdateToolCall:
      tool_call_id: call_W4t4W4TDhrZFiM9DxVox21PW
      entry:
        timestamp: ~
        entry_type:
          type: tool_use
          tool_name: Execute
          action_type:
            action: command_run
            command: "printf 'a' > \"/Users/tempuser/projects/vibe-kanban/a.txt\""
            result: ~
          status:
            status: failed
        content: "`printf 'a' > \"/Users/tempuser/projects/vibe-kanban/a.txt\"`"
        metadata: ~
  - AddEntry:
      timestamp: ~
      entry_type:
        type: assistant_message
      content: "Exec ended early: insufficient permission to proceed. Re-run with --auto medium or --auto high. For destructive commands, use --skip-permissions-unsafe."
      metadata: ~
  - AddEntry:
      timestamp: ~
      entry_type:
        type: error_message
      content: "Exec ended early: insufficient permission to proceed. Re-run with --auto medium or --auto high. For destructive commands, use --skip-permissions-unsafe."
      metadata: ~
event_count: 10
